data-pipeline/src/telemetry/mod.rs

```diff
@@ -0,0 +1,540 @@
+    pub async fn build(self) -> Result<TelemetryClient, TelemetryError> {
+        let (worker, handle) = TelemetryWorkerBuilder::new_fetch_host(
+            self.service_name.unwrap(),
+            self.language.unwrap(),
+            self.language_version.unwrap(),
+            self.tracer_version.unwrap(),
+        )
+        .spawn_with_config(self.config)
+        .await
+        .map_err(|_| TelemetryError::Builder)?;
+
+        Ok(TelemetryClient {
+            handle,
+            metrics: Metrics::new(&worker),
+            worker,
+        })
+    }
+}
+
+/// Telemetry handle used to send metrics to the agent
+pub struct TelemetryClient {
+    metrics: Metrics,
+    worker: TelemetryWorkerHandle,
+    handle: JoinHandle<()>,
+}
+
+impl TelemetryClient {
+    /// Sends metrics to the agent using a telemetry worker handle.
+    ///
+    /// # Arguments:
+    ///
+    /// * `telemetry_handle`: telemetry worker handle used to enqueue metrics.
+    pub fn send(&self, data: &SendDataResult) -> Result<(), TelemetryError> {
+        if data.requests_count > 0 {
+            let key = self.metrics.get(metrics::MetricKind::ApiRequest);
+            self.worker
+                .add_point(data.requests_count as f64, key, vec![])?;
+        }
+        if data.errors_network > 0 {
+            let key = self.metrics.get(metrics::MetricKind::ApiErrorsNetwork);
+            self.worker
+                .add_point(data.errors_network as f64, key, vec![])?;
+        }
+        if data.errors_timeout > 0 {
+            let key = self.metrics.get(metrics::MetricKind::ApiErrorsTimeout);
+            self.worker
+                .add_point(data.errors_timeout as f64, key, vec![])?;
+        }
+        if data.errors_status_code > 0 {
+            let key = self.metrics.get(metrics::MetricKind::ApiErrorsStatusCode);
+            self.worker
+                .add_point(data.errors_status_code as f64, key, vec![])?;
+        }
+        if data.bytes_sent > 0 {
+            let key = self.metrics.get(metrics::MetricKind::ApiBytes);
+            self.worker.add_point(data.bytes_sent as f64, key, vec![])?;
+        }
+        if data.chunks_sent > 0 {
+            let key = self.metrics.get(metrics::MetricKind::ChunksSent);
+            self.worker
+                .add_point(data.chunks_sent as f64, key, vec![])?;
+        }
+        if data.chunks_dropped > 0 {
+            let key = self.metrics.get(metrics::MetricKind::ChunksDropped);
+            self.worker
+                .add_point(data.chunks_dropped as f64, key, vec![])?;
+        }
+        if !data.responses_count_per_code.is_empty() {
+            let key = self.metrics.get(metrics::MetricKind::ApiResponses);
+            for (status_code, count) in &data.responses_count_per_code {
+                self.worker.add_point(
+                    *count as f64,
+                    key,
+                    vec![Tag::new("status_code", status_code.to_string().as_str()).unwrap()],
```

Comment:
Is unwrapping here ok? Should we be handling potential errors?

Reply:
Tag creation only fails if the tag name begins or ends with a `:` or is empty. That said there's little chances that this will panic but I think we need to avoid unwrap as much as possible in production code so I will refactor it.